### PR TITLE
Generalize SignedCheckpoint to take arbitrary Notes

### PR DESCRIPTION
**Pull request by asraa**
Generalizes SignedCheckpoint to a SignedNote:

SignedNote contains a textual representation of a `Note` and `note.Signatures`.
SignedCheckpoint uses a Checkpoint as the note, and marshals it as before into the `SignedNote.Note`.

This lays the groundwork for a SignedTimestamp which would contain a marshalled TimestampResponse as the Note.

For https://github.com/sigstore/rekor/issues/315

Signed-off-by: Asra Ali <asraa@google.com>

### CODE REVIEW

**Comment by asraa** [comment]
Merged main -- just looking for a signal on whether I'm missing something I can do in Go about specializing a type in an interface implementation

i.e. can i specify that a RekorSTH, which is a type with a SignedNote (containing a Note interface), always has a Checkpoint Note?

Maybe the best way is to write methods for RekorSTH to get the Hash/Size

**Comment by asraa** [comment]
@dekkagaijin if you have a few minutes could you please take a look? mostly on the comment above!

**Comment by bobcallaway** [comment]
Looking at `golang.org/x/mod/sumdb/note` it seems like they just leave it as a `string`. 

> Maybe the best way is to write methods for RekorSTH to get the Hash/Size

That would probably be where I'd start.

**Comment by asraa** [comment]
> Looking at golang.org/x/mod/sumdb/note it seems like they just leave it as a string.

Storing it as a text representation worked okay, won't need special methods to access Checkpoint data now.

File: pkg/util/signed_note.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+type SignedNote struct {
+	// Textual representation of a note to sign.
+	Note string
+	// Signatures are one or more signature lines covering the payload
+	Signatures []note.Signature
+}
+
+// Sign adds a signature to a SignedCheckpoint object
+// The signature is added to the signature array as well as being directly returned to the caller
+func (s *SignedNote) Sign(identity string, signer crypto.Signer, opts crypto.SignerOpts) (*note.Signature, error) {
```
Comment (dekkagaijin) [inline pkg/util/signed_note.go]: can replace this with `signature.Signer` from s/s to eliminate the hashing boilerplate

File: pkg/util/signed_note.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+		Base64: base64.StdEncoding.EncodeToString(sig),
+	}
+
+	s.Signatures = append(s.Signatures, signature)
+	return &signature, nil
+}
+
+// Verify checks that one of the signatures can be successfully verified using
+// the supplied public key
+func (s SignedNote) Verify(public crypto.PublicKey) bool {
```
Comment (dekkagaijin) [inline pkg/util/signed_note.go]: ditto this with `signature.Verifier`

**Comment by asraa** [comment]
Moved over to signature.Signer/Verifier. Sorry it took a second! Vacation + some bugs.

File: pkg/util/signed_note.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+	msg := []byte(s.Note)
+	digest := sha256.Sum256(msg)
+
+	for _, s := range s.Signatures {
+		sigBytes, err := base64.StdEncoding.DecodeString(s.Base64)
+		if err != nil {
+			return false
+		}
+		pk, _ := verifier.PublicKey()
```
Comment (dlorenc) [inline pkg/util/signed_note.go]: Do we need to check this error?

File: pkg/util/signed_note.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+		opts := []signature.VerifyOption{}
+		switch pk.(type) {
+		case *rsa.PublicKey:
```
Comment (dlorenc) [inline pkg/util/signed_note.go]: there's no fallthrough by default, I think you can do a multi-line case statement or use the explicit fallthrough operator.

File: pkg/util/signed_note.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+// String returns the String representation of the SignedNote
+func (s SignedNote) String() string {
+	var b strings.Builder
+	b.WriteString(s.Note)
+	b.WriteRune('\n')
+	for _, sig := range s.Signatures {
+		var hbuf [4]byte
+		binary.BigEndian.PutUint32(hbuf[:], sig.Hash)
+		sigBytes, _ := base64.StdEncoding.DecodeString(sig.Base64)
```
Comment (dlorenc) [inline pkg/util/signed_note.go:121]: Do we need to check this error?

**Review by dlorenc — COMMENTED** [review]
A few nits.

File: pkg/util/signed_note.go (hunk 3 above)
Comment (asraa) [reply pkg/util/signed_note.go]: I should, I was coding too transparently into the Verifiers (it never returns an error https://github.com/sigstore/sigstore/blob/a9aeb218f4d17318e29c0df3315c0496edf25a1a/pkg/signature/ecdsa.go#L138)

File: pkg/util/signed_note.go (hunk 5 above)
Comment (asraa) [reply pkg/util/signed_note.go:121]: Unmarshalling checks this, and creating a signature will guarantee this to be OK, so I think there's an implicit guarantee that any SignedNote is well-formed in  that respect. On the other hand, then I would probably make the field private. WDYT? (This is existing)

**Review by dlorenc — APPROVED** [review]

File: pkg/util/signed_note.go (hunk 5 above)
Comment (dlorenc) [reply pkg/util/signed_note.go:121]: Yeah this is probably fine.

**Pull request merged**